From dbb065489aa2d5a2099b6abb7298087dc94d7efd Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 2 Feb 2017 16:04:56 -0800
Subject: [PATCH] joyent/node-cueball#82 pools should error-out all outstanding
 claims when entering "failed" state

---
 lib/pool.js       | 7 +++++++
 test/pool.test.js | 9 +++++++++
 2 files changed, 16 insertions(+)

diff --git a/lib/pool.js b/lib/pool.js
index 9d11e01..8bcc13e 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -362,6 +362,13 @@ CueBallConnectionPool.prototype.state_failed = function (S) {
 	});
 
 	this._incrCounter('failed-state');
+
+	/* Fail all outstanding claims that are waiting for a connection. */
+	while (!this.p_waiters.isEmpty()) {
+		var hdl = this.p_waiters.shift();
+		if (hdl.isInState('waiting'))
+			hdl.fail(new mod_errors.PoolFailedError(self));
+	}
 };
 
 CueBallConnectionPool.prototype.state_running = function (S) {
diff --git a/test/pool.test.js b/test/pool.test.js
index 54b0e5b..0766a95 100644
--- a/test/pool.test.js
+++ b/test/pool.test.js
@@ -500,8 +500,17 @@ mod_tape.test('pool failure', function (t) {
 				index.b1[0].emit('error', new Error());
 				index.b1[1].emit('error', new Error());
 
+				var sawErr = false;
+				pool.claim(function (err) {
+					t.ok(err);
+					t.strictEqual(err.name,
+					    'PoolFailedError');
+					sawErr = true;
+				});
+
 				setTimeout(function () {
 					t.ok(pool.isInState('failed'));
+					t.ok(sawErr);
 
 					t.equal(connections.length, 1);
 					summarize();
-- 
2.21.0

